reana_job_controller/compute4punch_job_manager.py

```diff
@@ -0,0 +1,366 @@
+            (
+                f"requirements = {self.c4p_additional_requirements}"
+                if self.c4p_additional_requirements
+                else ""
+            ),
+            "queue 1",
+        ]
+        job_description = "\n".join(filter(None, job_description_template))
+
+        self.c4p_connection.exec_command(
+            f"cat <<< '{job_description}' > {self.job_description_path}"
+        )
+
+    def _create_c4p_job_execution_script(self) -> None:
+        """
+        Create job execution script for Compute4PUNCH
+        """
+        # The workflow workspace does not exist on Compute4PUNCH,
+        # therefore replace it with CONDOR_JOB_IWD
+        self.cmd = self.cmd.replace(self.workflow_workspace, "$_CONDOR_JOB_IWD")
+        logging.info(f"CMD is {self.cmd}")
+        self.cmd = self._encode_cmd(self.cmd)
+        job_execution_script_template = ["#!/bin/bash", f"{self.cmd}"]
+        job_execution_script = "\n".join(job_execution_script_template)
+
+        self.c4p_connection.exec_command(
+            f"cat <<< '{job_execution_script}' > {self.job_execution_script_path} && "
+            f"chmod +x {self.job_execution_script_path}"
+        )
+
+    def _create_c4p_workspace_environment(self) -> None:
+        """
+        Create workspace environment for REANA @ Compute4PUNCH
+        """
+        self.c4p_connection.exec_command(f"mkdir -p {self.c4p_abs_workspace_path}")
+        self.c4p_connection.exec_command(
+            os.path.join(
+                f"mkdir -p {os.path.join(self.c4p_abs_workspace_path, 'logs')}"
+            )
+        )
+
+    @classmethod
+    def _download_output_directory(
+        cls, sftp_client: SFTPClient, remote_dir: str, local_dir: str
+    ) -> None:
+        """
+        Download output directory and content to a local directory.
+
+        :param sftp_client: SFTP client to use for downloading
+        :type sftp_client: SFTPClient
+        :param remote_dir: Remote directory to download
+        :type remote_dir: str
+        :param local_dir: Local destination directory
+        :type local_dir: str
+        """
+        skipped_output_files = (
+            "cluster.log",
+            "_condor_stdout",
+            "_condor_stderr",
+            "run.sh",
+            "submit.jdl",
+            "tmp",
+            "var_tmp",
+        )  # skip intermediate and temporary htcondor files
+        os.path.exists(local_dir) or os.makedirs(local_dir)
+        for item in filter(
+            lambda x: x.filename not in skipped_output_files,
+            sftp_client.listdir_attr(remote_dir),
+        ):
+            remote_path = os.path.join(remote_dir, item.filename)
+            local_path = os.path.join(local_dir, item.filename)
+            if S_ISDIR(item.st_mode):
+                cls._download_output_directory(sftp_client, remote_path, local_path)
+            else:
+                sftp_client.get(remote_path, local_path)
```

Comment:
Are we always sure that parent directories are always returned before the files/directories that are contained in them? Or does `sftp_client.get` create the parent directories if missing?

Reply:
In this case we are sure. I am doing a basically a `listdir` in the parent directory and recursively call the same function for all sub-directories. So, the directory will be always created, before the sub-directory is actually traversed.